### PR TITLE
[IOS-351]: Enhanced logging

### PR DESCRIPTION
**Pull request by yuliaveres**
Add custom events to Crashlytics

### CODE REVIEW

File: Toshi/APIClients/CrashlyticsClient.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+@objc final class CrashlyticsLogger: NSObject {
+
+    @objc static func log(_ string: String, attributes: [String: Any]? = nil) {
+        CLSLogv("%@", getVaList([string]))
+
+        var resultAttributes: [String: Any] = [:]
+
+        if let userID = Cereal.shared.address as String? {
```
Comment (roberthein) [inline Toshi/APIClients/CrashlyticsClient.swift]: This casting `as String?` makes something that isn't an Optional into an Optional so you have to unwrap. But if you don't cast it you also don't have to unwrap it. 🤷‍♂️

File: Toshi/APIClients/CrashlyticsClient.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            resultAttributes["user_id"] = userID
+        }
+
+        if let passedAttributes = attributes as [String: Any]? {
```
Comment (roberthein) [inline Toshi/APIClients/CrashlyticsClient.swift]: This casting `as [String: Any]?` does not do anything and can be removed.

File: Toshi/APIClients/CrashlyticsClient.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            for (key, value) in passedAttributes {
+                resultAttributes[key] = value
+            }
+        }
+
+        Answers.logCustomEvent(withName: string,
+                               customAttributes: resultAttributes)
```
Comment (roberthein) [inline Toshi/APIClients/CrashlyticsClient.swift]: Misplaced newline?

File: Toshi/APIClients/CrashlyticsClient.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    }
+}
```
Comment (roberthein) [inline Toshi/APIClients/CrashlyticsClient.swift]: You can also simplify this function and write it like this:
```
@objc static func log(_ string: String, attributes: [String: Any]? = nil) {
    CLSLogv("%@", getVaList([string]))

    var resultAttributes: [String: Any] = ["user_id": Cereal.shared.address]
    attributes?.forEach { key, value in resultAttributes[key] = value }

    Answers.logCustomEvent(withName: string, customAttributes: resultAttributes)
}
```
What do you think?

File: Toshi/Controllers/Favorites/ContactController.swift [hunk 5]
```diff
@@ -568,6 +570,7 @@ extension ContactController: PaymentControllerDelegate {
                     strongSelf.messageSender?.send(outgoingMessage, success: {
                         print("message sent")
                     }, failure: { error in
+                        CrashlyticsLogger.log("Can not send message", attributes: ["Error": error.localizedDescription])
```
Comment (roberthein) [inline Toshi/Controllers/Favorites/ContactController.swift]: In the attributes I see `error` and `Error` and 7 times `occured`, can you maybe make an `enum` for this? 🙏

Comment (roberthein) [reply Toshi/Controllers/Favorites/ContactController.swift]: Or you can do something like this:
```
typealias LogAttributeType = String

extension LogAttributeType {
    
    static var error: String {
        return "error"
    }
    
    static var occured: String {
        return "occured"
    }
}
```
Then in the `log` function replace all `[String: Any]` with `[LogAttributeType: Any]` and then you can call `log` like this:
```
CrashlyticsLogger.log("No current user during session", attributes: [.occured: "Profile edit item"])
```
What do you think?

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 3 above)
Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: yup

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 1 above)
Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: wdym `makes something that isn't an Optional into an Optional` ? address is optional, depends on current session ..

Comment (roberthein) [reply Toshi/APIClients/CrashlyticsClient.swift]: `Cereal.shared.address` is not Optional.

Comment (roberthein) [reply Toshi/APIClients/CrashlyticsClient.swift]: Or am I missing something?

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 2 above)
Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: here agree :)

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 4 above)
Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: i dnt want to crash app while trying to catch unexpected event :D
so if that happens f.e. during session start and Cereal is not setup yet, it does not have any address, i can't pass it into dictionary like it is suggested 🤷‍♂️ 

great suggestion regarding attributes though 👍

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 1 above)
Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: oh, i meant NOT OPTIONAL sorry. so i need this cast here

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 4 above)
Comment (roberthein) [reply Toshi/APIClients/CrashlyticsClient.swift]: This `Cereal.shared.address as String?` does not save you from a crash.

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 1 above)
Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: sorry, on master it is not optional right ! i am living in the context of multisession now :(
but we are targeting this. let support it here from now on ? or should i be back here after multisession is merged and change it back to how it is now ?

Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: as soon as multisession is there, all cereal data is optional and depends on user session

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 4 above)
Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: yes, i mean cereal being setup for user only . so i mean address being optional (very soon)

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 1 above)
Comment (roberthein) [reply Toshi/APIClients/CrashlyticsClient.swift]: This cast is wrong even if `Cereal.shared.address` was optional.

Comment (roberthein) [reply Toshi/APIClients/CrashlyticsClient.swift]: When something is optional, you unwrap it. No need for casting.

Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: can you explain ?

Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: got it

Comment (yuliaveres) [reply Toshi/APIClients/CrashlyticsClient.swift]: now it is impossible to unwrap though . ok, will be back here later

Comment (roberthein) [reply Toshi/APIClients/CrashlyticsClient.swift]: Yeah so, when `Cereal.shared.address` will be optional in the future, you should do:
```
if let userID = Cereal.shared.address {
    resultAttributes["user_id"] = userID
}
```
No casting.
Now it's not optional so you can use `Cereal.shared.address ` without casting or unwrapping.

File: Toshi/APIClients/CrashlyticsClient.swift (hunk 4 above)
Comment (roberthein) [reply Toshi/APIClients/CrashlyticsClient.swift]: Even if it's optional, you still don't have to cast it.

File: Toshi/Controllers/Favorites/ContactController.swift (hunk 5 above)
Comment (yuliaveres) [reply Toshi/Controllers/Favorites/ContactController.swift]: i like the idea

Comment (yuliaveres) [reply Toshi/Controllers/Favorites/ContactController.swift]: but i am wondering if `LogAttributeType ` is smth we want. lets assume we are gping to use keys strings in other places of the app. and also would like to name keys in proper way. creating another typealias for `String` is not a good idea i guess. maybe we should create one for global use.. like "Key", "KeyTitle" .. ? any ideas ?

Comment (roberthein) [reply Toshi/Controllers/Favorites/ContactController.swift]: Yeah a `typealias` on `String`  may not be the best idea. So what do you propose for the `KeyTitle`? A `struct`?

Comment (yuliaveres) [reply Toshi/Controllers/Favorites/ContactController.swift]: what i meant is just other name for typealias to have it more generic 🙈  😄

Comment (roberthein) [reply Toshi/Controllers/Favorites/ContactController.swift]: Aah yes, in that case a `typealias` makes more sense yes. 👍

Comment (yuliaveres) [reply Toshi/Controllers/Favorites/ContactController.swift]: using struct here is not really possible for what we need. it should be `hashable` (cause used as key in dictionary) + it should operate on instance with properties. that is not need for string case .. ?

**Pull request merged**